File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/dao/OcgmRepository.java
package uk.gov.justice.digital.hmpps.riskprofiler.dao;

import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Repository;
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.Ocgm;
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.OcgmList;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

import static uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.FileType.OCGM;

@Repository
@Slf4j
public class OcgmRepository implements DataRepository<OcgmList> {

    private final ImportedFile<OcgmList> data = new ImportedFile<>();

    @Override
    public void process(List<List<String>> csvData, final String filename, final LocalDateTime timestamp) {

        data.setFileTimestamp(timestamp);
        data.setFileName(filename);
        data.setFileType(OCGM);
        data.reset();

        csvData.stream().filter(p -> data.getIndex().getAndIncrement() > 0)
                .forEach(p -> {
                    try {
                        final var key = p.get(Ocgm.NOMIS_ID_POSITION);
                        if (StringUtils.isNotBlank(key)) {
                            if (!NOMS_ID_REGEX.matcher(key).matches()) {
                                log.warn("Invalid Key in line {} for Key {}", data.getIndex().get(), key);
                                data.getLinesInvalid().incrementAndGet();
                            } else {
                                var ocgId = p.get(Ocgm.OCG_ID_POSITION);
                                if (StringUtils.isBlank(ocgId)) {
                                    log.warn("No OCG Id in line {} for Key {}", data.getIndex().get(), key);
                                    data.getLinesInvalid().incrementAndGet();
                                } else {
                                    var ocgmLine = Ocgm.builder()
                                            .nomisId(key)
                                            .ocgId(StringUtils.trimToNull(ocgId))
                                            .standingWithinOcg(StringUtils.trimToNull(p.get(Ocgm.STANDING_POSITION)))
                                            .build();

                                    var dataSet = data.getDataSet().get(key);
                                    if (dataSet != null) {
                                        dataSet.getData().add(ocgmLine);
                                    } else {
                                        data.getDataSet().put(key, OcgmList.builder().nomisId(key).ocgm(ocgmLine).build());
                                    }
                                    data.getLinesProcessed().incrementAndGet();
                                }
                            }
                        } else {
                            log.warn("Missing key in line {}", data.getIndex().get(), key);
                            data.getLinesInvalid().incrementAndGet();
                        }
                    } catch (Exception e) {
                        log.warn("Error in Line {}", data.getIndex().get(), p);
                        data.getLinesError().incrementAndGet();
                    }
                });
        log.info("Lines total {}, processed {}, dups {}, invalid {}, errors {}", data.getIndex().get(),
                data.getLinesProcessed().get(), data.getLinesDup().get(), data.getLinesInvalid().get(), data.getLinesError().get());


    }

    public ImportedFile<OcgmList> getData() {
        return data;
    }

    public LocalDateTime getFileTimestamp() {
        return data.getFileTimestamp();
    }

    @Override
    public Optional<OcgmList> getByKey(String key) {
        return Optional.ofNullable(data.getDataSet().get(key));
    }

}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/dao/OcgmRepository.kt
package uk.gov.justice.digital.hmpps.riskprofiler.dao

import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Repository
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.FileType
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.Ocgm
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.OcgmList
import java.time.LocalDateTime

@Repository
class OcgmRepository : DataRepository<OcgmList>() {
  override fun doProcess(
    csvData: List<List<String>>,
    filename: String,
    timestamp: LocalDateTime,
    data: ImportedFile<OcgmList>,
  ) {
    data.fileTimestamp = timestamp
    data.fileName = filename
    data.fileType = FileType.OCGM
    data.reset()
    csvData.stream().filter { data.index.getAndIncrement() > 0 }
      .forEach { p ->
        try {
          val key = p[Ocgm.NOMIS_ID_POSITION]
          if (StringUtils.isNotBlank(key)) {
            if (!NOMS_ID_REGEX.matcher(key).matches()) {
              log.warn("Invalid Key in line {} for Key {}", data.index.get(), key)
              data.linesInvalid.incrementAndGet()
            } else {
              val ocgId = p[Ocgm.OCG_ID_POSITION]
              if (StringUtils.isBlank(ocgId)) {
                log.warn("No OCG Id in line {} for Key {}", data.index.get(), key)
                data.linesInvalid.incrementAndGet()
              } else {
                val ocgmLine =
                  Ocgm(key, StringUtils.trimToNull(ocgId), StringUtils.trimToNull(p[Ocgm.STANDING_POSITION]))
                val dataSet = data.dataSet!![key]

                if (dataSet != null) {
                  dataSet.data.add(ocgmLine)
                } else {
                  data.dataSet!![key] = OcgmList(key, ocgmLine, null)
                }

                data.linesProcessed.incrementAndGet()
              }
            }
          } else {
            log.warn("Missing key in line {} key [{}]", data.index.get(), key)
            data.linesInvalid.incrementAndGet()
          }
        } catch (e: Exception) {
          log.warn("Error in Line {} data [{}]", data.index.get(), p)
          data.linesError.incrementAndGet()
        }
      }
    log.info(
      "Lines total {}, processed {}, dups {}, invalid {}, errors {}",
      data.index.get(),
      data.linesProcessed.get(),
      data.linesDup.get(),
      data.linesInvalid.get(),
      data.linesError.get(),
    )
  }

  companion object {
    private val log = LoggerFactory.getLogger(OcgmRepository::class.java)
  }
}
